### PR TITLE
Cherry pick makefile & rust-toolchain changes

### PR DESCRIPTION
**Pull request by breeswish**
Reverts tikv/tikv#4986
Repeats https://github.com/tikv/tikv/pull/4963

Previously we need to release GA version, so the compiler change was reverted. Now GA is released, so we can move forward to bring compiler changes again.

This PR cherry picks some recent changes about the Makefile and rust-toolchain, so that our build cache won't be invalidated when switching to release-3.0 branch from master branch.

### CODE REVIEW

**Comment by BusyJay** [comment]
Do we do dist_release on release-3.0 branch?

/cc @zhouqiang-cl

**Comment by zhouqiang-cl** [comment]
@BusyJay yes

File: Dockerfile [hunk 1]
```diff
@@ -14,6 +14,9 @@ RUN sed -i '/fuzz/d' Cargo.toml && \
     sed -i '/test\_/d' Cargo.toml && \
     sed -i '/profiler/d' Cargo.toml
 
+# Use Makefile to build
+COPY Makefile ./
```
Comment (overvenus) [inline Dockerfile]: Should we pick https://github.com/tikv/tikv/pull/5017?

Comment (breeswish) [reply Dockerfile]: Nice catch

**Pull request merged**